fix CompositeImplicitAutograd ops improperly labeled

bdhirsh opened the pull request:
I checked `derivatives.yaml`, and it doesn't look like `logical_not/and/xor` are meant to work with autograd. Those 3 ops are currently set as `CompositeImplicitAutograd` though, implying that they do work with autograd. Updating them to be CompositeExplicitAutograd instead.

This came up because I'm trying to improve the error checking in external backend codegen, and these ops being improperly labeled incorrectly triggers my new error checks for XLA (see https://github.com/pytorch/pytorch/pull/67090)

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#69169 fix CompositeImplicitAutograd ops improperly labeled**

Differential Revision: [D32739976](https://our.internmc.facebook.com/intern/diff/D32739976)

bdhirsh committed:
fix CompositeImplicitAutograd ops improperly labeled

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32736964).

zou3519 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -922,10 +922,14 @@
   device_check: NoCheck   # TensorIterator
   variants: function, method
+  dispatch:
+    CompositeExplicitAutograd: logical_not
```
You also need to add a backward formula that says the input is not differentiable, right?

e.g. https://github.com/pytorch/pytorch/blob/aa2163eba53036d998d564138e10f18634c890b9/tools/autograd/derivatives.yaml#L2417-L2419

bdhirsh replied:
ah, right. thanks!

zou3519 replied:
I feel like we should do this by default, this is a common thing folks miss

soulitzer replied:
Do you mean to add this because we don't want to raise an error if the inputs require grad? Is it possible for it to not raise an error here even without the formula, because the output is bool... i.e, the number of differentiable outputs is 0.

zou3519 replied:
Without this PR, logical_not does not raise an error when passed inputs that require_grad:
```
>>> import torch
>>> x = torch.randn(3, requires_grad=True)
>>> torch.logical_not(x)
tensor([False, False, False])
```
So we probably want to replicate that behavior, and to replicate that behavior the way I know of is to add the backward formula (otherwise, the operator errors out, right?)

soulitzer replied:
Ah yeah you are definitely right here. Codegen has no way of checking the dtype of the output, so it can only rely on output_differentiability, which can only exist if the formula exists. More generally though the logic is slightly more involved.
For example `new_zeros` is composite, but doesn't have a formula, and shouldn't error because it is listed in `DONT_REQUIRE_DERIVATIVE`.

zou3519 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -961,10 +973,14 @@
 - func: logical_or_(Tensor(a!) self, Tensor other) -> Tensor(a!)
   device_check: NoCheck   # TensorIterator
   variants: method
+  dispatch:
+    CompositeExplicitAutograd: logical_or_
```
This PR fixes the root cause of the composite compliance test failure for these operators (https://github.com/pytorch/pytorch/blob/aa2163eba53036d998d564138e10f18634c890b9/torch/testing/_internal/common_methods_invocations.py#L9707-L9708). You'll probably have to remove those expected failures

bdhirsh committed:
Update on "fix CompositeImplicitAutograd ops improperly labeled"

I checked `derivatives.yaml`, and it doesn't look like `logical_not/and/xor` are meant to work with autograd. Those 3 ops are currently set as `CompositeImplicitAutograd` though, implying that they do work with autograd. Updating them to be CompositeExplicitAutograd instead.

This came up because I'm trying to improve the error checking in external backend codegen, and these ops being improperly labeled incorrectly triggers my new error checks for XLA (see https://github.com/pytorch/pytorch/pull/67090)




[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32739976).

bdhirsh commented on tools/autograd/derivatives.yaml:
```diff
@@ -1727,6 +1727,15 @@
+  output_differentiability: [False]
+
+- name: logical_and(Tensor self, Tensor other) -> Tensor
+  output_differentiability: [False]
```
@soulitzer what's the best way to declare these ops as non-differentiable? I guess the best case scenario would be if we can error out immediately in the forward pass, if someone tries to call `torch.logical_and(torch.ones(2, requres_grad=True))`.

I'm getting codegen errors that say `RuntimeError: output_differentiability=False for inplace operation (version_counter won't get updated)`. I see another option from the docs is to add these ops to the `DONT_REQUIRE_DERIVATIVE` in `gen_variable_type`?

soulitzer replied:
It is pretty counterintuitive but declaring these ops as non-differentiable actually makes it so that they do _not_ error when someone passes in a tensor that has `requires_grad=True`. The default behavior when no formula is specified is actually to error immediately in the forward pass.

A good example of why that is the case is `eq`, we wouldn't want users to have to `detach` everytime they want to compare two tensors. Maybe it is less clear what we really want to do in the `logical_{and,or}` cases, but adding it would be good to avoid bc-breaking issues.

I'm actually wondering how outdated that error is, since we update the version counter in the ADInplaceOrView kernel now. Perhaps we should just remove it?

It is also weird that `eq` does a similar thing (i.e., it has a in-place variant), but doesn't run into the error. I'm not sure why `eq` is able to work, but if you look into that perhaps you can emulate how it is done there.

Adding to `DONT_REQUIRE_DERIVATIVE` should also work, so I guess we can do that as a easy workaround for now.

bdhirsh replied:
> A good example of why that is the case is eq, we wouldn't want users to have to detach everytime they want to compare two tensors.

Ah okay, then keeping the current behavior of not immediately erroring out in the forward pass sounds reasonable, to not be BC-breaking.

> Adding to DONT_REQUIRE_DERIVATIVE should also work, so I guess we can do that as a easy workaround for now.

I'll try this for now, thanks! I looked at `eq` and I don't think I'm doing anything differently for `logical_and/xor/not`.

bdhirsh committed:
Update on "fix CompositeImplicitAutograd ops improperly labeled"

I checked `derivatives.yaml`, and it doesn't look like `logical_not/and/xor` are meant to work with autograd. Those 3 ops are currently set as `CompositeImplicitAutograd` though, implying that they do work with autograd. Updating them to be CompositeExplicitAutograd instead.

This came up because I'm trying to improve the error checking in external backend codegen, and these ops being improperly labeled incorrectly triggers my new error checks for XLA (see https://github.com/pytorch/pytorch/pull/67090)


Differential Revision: [D32739976](https://our.internmc.facebook.com/intern/diff/D32739976)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32739976).

bdhirsh committed:
Update on "fix CompositeImplicitAutograd ops improperly labeled"

I checked `derivatives.yaml`, and it doesn't look like `logical_not/and/xor` are meant to work with autograd. Those 3 ops are currently set as `CompositeImplicitAutograd` though, implying that they do work with autograd. Updating them to be CompositeExplicitAutograd instead.

This came up because I'm trying to improve the error checking in external backend codegen, and these ops being improperly labeled incorrectly triggers my new error checks for XLA (see https://github.com/pytorch/pytorch/pull/67090)


Differential Revision: [D32739976](https://our.internmc.facebook.com/intern/diff/D32739976)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32739976).

bdhirsh committed:
Update on "fix CompositeImplicitAutograd ops improperly labeled"

I checked `derivatives.yaml`, and it doesn't look like `logical_not/and/xor` are meant to work with autograd. Those 3 ops are currently set as `CompositeImplicitAutograd` though, implying that they do work with autograd. Updating them to be CompositeExplicitAutograd instead.

This came up because I'm trying to improve the error checking in external backend codegen, and these ops being improperly labeled incorrectly triggers my new error checks for XLA (see https://github.com/pytorch/pytorch/pull/67090)


Differential Revision: [D32739976](https://our.internmc.facebook.com/intern/diff/D32739976)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32739976).

bdhirsh commented:
@soulitzer @zou3519 I think this is ready for another round of review

bdhirsh committed:
Update on "fix CompositeImplicitAutograd ops improperly labeled"

I checked `derivatives.yaml`, and it doesn't look like `logical_not/and/xor` are meant to work with autograd. Those 3 ops are currently set as `CompositeImplicitAutograd` though, implying that they do work with autograd. Updating them to be CompositeExplicitAutograd instead.

This came up because I'm trying to improve the error checking in external backend codegen, and these ops being improperly labeled incorrectly triggers my new error checks for XLA (see https://github.com/pytorch/pytorch/pull/67090)


Differential Revision: [D32739976](https://our.internmc.facebook.com/intern/diff/D32739976)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32739976).

soulitzer reviewed (approved):
LGTM
